packages/daemon/src/mail.js

```diff
@@ -331,93 +246,47 @@ export const makeMailboxMaker = ({
-        recipientId,
-      );
-      const recipientInternal = await recipientController.internal;
-      if (recipientInternal === undefined || recipientInternal === null) {
-        throw new Error(
-          `panic: a receive request function must exist for every agent`,
-        );
-      }
-
-      // @ts-expect-error We sufficiently check if recipientInternal or deliverToRecipient is undefined
-      const { respond: deliverToRecipient } = recipientInternal;
-      if (deliverToRecipient === undefined) {
-        throw new Error(
-          `panic: a receive request function must exist for every agent`,
-        );
-      }
-
+    const request = async (toName, description, responseName) => {
       if (responseName !== undefined) {
-        const responseP = responses.get(responseName);
-        if (responseP !== undefined) {
-          return responseP;
+        const responseId = petStore.identifyLocal(responseName);
+        if (responseId !== undefined) {
+          return provide(responseId);
         }
       }
 
-      // Note: consider sending to each mailbox with different powers.
-      // Behold, recursion:
-      // eslint-disable-next-line
-      const recipientResponseP = deliverToRecipient(
-        what,
-        responseName,
-        selfId,
-        petStore,
-      );
-      // Send to own inbox.
-      const selfResponseP = respond(
-        what,
-        responseName,
+      const toId = petStore.identifyLocal(toName);
+      if (toId === undefined) {
+        throw new Error(`Unknown recipient ${toName}`);
+      }
+      const to = await provideAgentForHandle(toId);
+
+      const { request: req, response: responseIdP } = makeRequest(
+        description,
         selfId,
-        petStore,
-        // Sender expects the handle formula identifier.
-        recipientId,
+        toId,
       );
-      const newResponseP = Promise.race([recipientResponseP, selfResponseP]);
 
-      if (responseName !== undefined) {
-        responses.set(responseName, newResponseP);
-      }
+      // Note: consider sending to each mailbox with different powers.
+      E(to).deliver(req);
+      // Send to own inbox.
+      deliver(req);
 
-      return newResponseP;
-    };
+      const responseId = await responseIdP;
+      const responseP = provide(responseId);
 
-    /** @type {import('./types.js').PetStore['rename']} */
-    const rename = async (fromName, toName) => {
-      await petStore.rename(fromName, toName);
-      const id = responses.get(fromName);
-      if (id !== undefined) {
-        responses.set(toName, id);
-        responses.delete(fromName);
+      if (responseName !== undefined) {
+        await petStore.write(responseName, responseId);
       }
-    };
-
-    /** @type {import('./types.js').PetStore['remove']} */
-    const remove = async petName => {
-      await petStore.remove(petName);
-      responses.delete(petName);
-    };
```

Comment:
In the new framing, the `responses` memo turns out to be duplicative. Since we all see the same identifiers now, we can lean on the memo beneath `provide`. Because we don’t need to coördinate caches, the mailbox no longer needs to intermediate these pet store methods. Concerns: separated.